test/infrastructure/docker/internal/controllers/backends/inmemory/inmemorymachine_backend.go

```diff
@@ -36,147 +35,34 @@ import (
 }
 
-// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=inmemorymachines,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=inmemorymachines/status;inmemorymachines/finalizers,verbs=get;update;patch
-// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;machinesets;machines,verbs=get;list;watch
-// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
-
-// Reconcile handles InMemoryMachine events.
-func (r *InMemoryMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
-	// Fetch the InMemoryMachine instance
-	inMemoryMachine := &infrav1.InMemoryMachine{}
-	if err := r.Client.Get(ctx, req.NamespacedName, inMemoryMachine); err != nil {
-		if apierrors.IsNotFound(err) {
-			return ctrl.Result{}, nil
-		}
-		return ctrl.Result{}, err
-	}
-
-	// Add finalizer first if not set to avoid the race condition between init and delete.
-	if finalizerAdded, err := finalizers.EnsureFinalizer(ctx, r.Client, inMemoryMachine, infrav1.MachineFinalizer); err != nil || finalizerAdded {
-		return ctrl.Result{}, err
-	}
-
-	// AddOwners adds the owners of InMemoryMachine as k/v pairs to the logger.
-	// Specifically, it will add KubeadmControlPlane, MachineSet and MachineDeployment.
-	ctx, log, err := clog.AddOwners(ctx, r.Client, inMemoryMachine)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-
-	// Fetch the Machine.
-	machine, err := util.GetOwnerMachine(ctx, r.Client, inMemoryMachine.ObjectMeta)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-	if machine == nil {
-		log.Info("Waiting for Machine Controller to set OwnerRef on InMemoryMachine")
-		return ctrl.Result{}, nil
-	}
-
-	log = log.WithValues("Machine", klog.KObj(machine))
-	ctx = ctrl.LoggerInto(ctx, log)
-
-	// Fetch the Cluster.
-	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
-	if err != nil {
-		log.Info("InMemoryMachine owner Machine is missing cluster label or cluster does not exist")
-		return ctrl.Result{}, err
-	}
-	if cluster == nil {
-		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterNameLabel))
-		return ctrl.Result{}, nil
-	}
-
-	log = log.WithValues("Cluster", klog.KObj(cluster))
-	ctx = ctrl.LoggerInto(ctx, log)
-
-	if isPaused, conditionChanged, err := paused.EnsurePausedCondition(ctx, r.Client, cluster, inMemoryMachine); err != nil || isPaused || conditionChanged {
-		return ctrl.Result{}, err
+// ReconcileNormal handle in memory backend for DevMachine not yet deleted.
+func (r *MachineBackendReconciler) ReconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, _ *infrav1.DevCluster, machine *clusterv1.Machine, inMemoryMachine *infrav1.DevMachine) (ctrl.Result, error) {
+	if inMemoryMachine.Spec.Backend.InMemory == nil {
+		panic("MachineBackendReconciler can't be called for DevMachines without an InMemory backend")
```

Comment:
[Applies to all panics]

Why we are using panic instead of return error?

(CR will catch the panics and log them anyway)

Reply:
I have a general tendency to only use panic if absolutely necessary